central/src/gitlab.rs

```diff
@@ -0,0 +1,105 @@
+    #[clap(long, env)]
+    pub gitlab_api_access_token: String,
+}
+
+#[derive(Deserialize)]
+struct GitLabApiReponseBody {
+    token: String,
+}
+
+pub struct GitLabProjectAccessTokenProvider {
+    config: GitLabApiConfig,
+    client: reqwest::Client,
+}
+
+impl GitLabProjectAccessTokenProvider {
+    pub fn try_init() -> Option<Self> {
+        match GitLabApiConfig::try_parse() {
+            Ok(config) => Some(Self {
+                config,
+                client: reqwest::Client::new(),
+            }),
+            Err(e) => {
+                println!("{e}");
+                None
+            }
+        }
+    }
+
+    /// Create a project access token using the GitLab API
+    pub async fn create_token(
+        &self,
+        name: &str,
+        client_config: GitLabProjectAccessTokenConfig,
+    ) -> Result<SecretResult, String> {
+        // Expire in 1 week
+        let expires_at = (chrono::Local::now() + chrono::TimeDelta::weeks(1))
+            .format("%Y-%m-%d")
+            .to_string();
+
+        let response = self
+            .client
+            .post(format!(
+                "{}api/v4/projects/{}/access_tokens",
+                self.config.gitlab_url,
+                urlencoding::encode(&client_config.project_path)
+            ))
+            .header("PRIVATE-TOKEN", &self.config.gitlab_api_access_token)
+            .json(&json!({
+                "name": format!("secret-sync-{name}"),
+                // The "read_repository" scope is required for git clone/pull permissions
+                "scopes": ["read_repository"],
+                // Access level 20 (Reporter) is the lowest level that allows git clone/pull
+                "access_level": 20,
+                "expires_at": expires_at,
+            }))
+            .send()
+            .await
+            .map_err(|e| e.to_string())?;
+
+        if response.status().is_success() {
+            let body: GitLabApiReponseBody = response.json().await.map_err(|e| e.to_string())?;
+            Ok(SecretResult::Created(body.token))
+        } else {
+            Err(format!(
+                "HTTP status error {} for url {} with body {}",
+                response.status(),
+                response.url().clone(),
+                response.text().await.unwrap()
+            ))
+        }
+    }
+
+    /// Simulate a git fetch to check the validity of the token
+    pub async fn validate_token(
+        &self,
+        _name: &str,
+        secret: &str,
+        client_config: &GitLabProjectAccessTokenConfig,
+    ) -> Result<bool, String> {
+        let response = self
+            .client
+            .get(format!(
+                "{}{}.git/info/refs?service=git-upload-pack",
+                self.config.gitlab_url, &client_config.project_path
+            ))
+            .basic_auth("any-non-blank-username", Some(secret))
```

Comment:
```suggestion
            .basic_auth("samply-secret-sync", Some(secret))
```

Reply:
`"any-non-bank-username"` should hopefully convey that any string would work. `"samply-secret-sync"` could mislead the reader into thinking that there is a user with that name of GitLab. Thoughts?

Reply:
Fair point, new suggestion:
```suggestion
            .basic_auth("placeholder-for-samply-secret-sync", Some(secret))
```